adagrams/game.py

```diff
@@ -1,11 +1,157 @@
+}
+
+
 def draw_letters():
-    pass
+# Create a list of letters using their frequencies
+# pull letters from the list using random.randint
+# create a new list from 10 random letters
+    list_letters = []   
+
+    for letter, freq in LETTER_POOL.items():
+        for i in range (0,freq):
+            list_letters.append(letter)
+
+    #return list_letters
+
+    hand = []
+    b = len(list_letters)
+    while len(hand)<10:
+        index  = random.randint(0,b-1)
+        new_letter = list_letters[index]
+        if hand.count(new_letter) < LETTER_POOL[new_letter]:
+            hand.append(new_letter)
+    return hand
+
+# Make all letter in word Capital
+# function returns True or False
 
 def uses_available_letters(word, letter_bank):
-    pass
 
-def score_word(word):
-    pass
+    #letter_bank = draw_letters()
+    copy_letter_bank = letter_bank[:]
+    word_up = word.upper()
+    
+    for letter in word_up:
+        if letter not in copy_letter_bank or word_up.count(letter) > copy_letter_bank.count(letter):
+            return False
+    return True
+
+
+# create a dictionary with points for each letter
+# create new variable total_score, assign 0 to it
+# for each letter add corresponding points from score dictionary
+# if len(word) >= 7 add 8 points
+# function returns score 
 
+def score_word(word):
+    total_score = 0
+    word_up = word.upper()
+    for letter in word_up:
+        total_score += score_dict[letter]
+    if len(word_up) >= 7 and len(word_up)<= 10:
+        total_score += 8
+    return total_score 
+    
+# create a dictionary with key as a word and value as a score
+# iterate through the dictionary using items() method to get max_score
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    words_score_dict = {}
+    for word in word_list:
+        words_score_dict[word] = score_word(word)
+    
+# Look for max score, create variable max_score
+# create a list of words with max score in case there are several words with same score
+    max_score = 0
+    max_score_words = []
+    for word, w_score in words_score_dict.items():
+        if w_score > max_score:
+            max_score = w_score
+            max_score_words = [word]
+        elif w_score == max_score:
+            max_score_words.append(word)
+# print(max_score_words)        
+# if there's one word with max score  
+
+    if len(max_score_words) == 1:
+        return (max_score_words[0], max_score) 
+    
+ # the case when there are several words with the same score   
+    elif len(max_score_words) > 1:
+
+        best_score_word = max_score_words[0]
+    
+        for word in max_score_words:
+            if len(word) == 10:
+                best_score_word = word
+                return (best_score_word, max_score) 
+           
+            elif len(word)<len(best_score_word):
+                best_score_word = word
+                return (best_score_word, max_score) 
+        return (best_score_word, max_score)    
```

Comment:
Overall, this code works fine, and separating everything out into individual dictionaries and lists can really help when making a first pass. You get to see each step and break out the logic quite a bit! When it comes to refactoring, however, it is possible to decrease the space complexity while maintaining similar time complexity. 

With an algorithm like this, we don't need to store every score and word, just the highest. If you would like to refactor, see if you can use the same logic without holding on to every piece of data. Figuring out what we DON'T need can be just as important as what we do! Specifically, we don't need to hold onto all the words with the max score. We can make the comparisons as we score each word in the list!